src/electionguard/async.hpp

```diff
@@ -0,0 +1,324 @@
+            _condition.wait(lock, [this] { return _count != 0; });
+            --_count;
+        }
+        void unlock()
+        {
+            unique_lock<mutex> lock(_mutex);
+            ++_count;
+            _condition.notify_one();
+        }
+
+        void wait() { std::scoped_lock _lock(*this); }
+
+      private:
+        mutex _mutex;
+        condition_variable _condition;
+        atomic<size_t> _count;
+    };
+
+    template <typename T> EG_INTERNAL_API bool isReady(const future<T> &f)
+    {
+        if (f.valid()) {
+            return f.wait_for(seconds(0)) == future_status::ready;
+        } else {
+            return false;
+        }
+    }
+
+    template <typename T>
+    EG_INTERNAL_API vector<unique_ptr<T>> wait_all(vector<future<unique_ptr<T>>> &tasks)
+    {
+        vector<unique_ptr<T>> results;
+        do {
+            for (auto &task : tasks) {
+                if (isReady(task)) {
+                    results.push_back(std::move(task.get()));
+                }
+            }
+        } while (results.size() < tasks.size());
+        return results;
+    }
+
+    /// <summary>
+    /// A simple asynchronous thread safe queue using locks.
+    /// </summary>
+    template <typename T> class EG_INTERNAL_API AsyncQueue
+    {
+      public:
+        AsyncQueue() : _head(new Node), _tail(_head.get()) {}
+        AsyncQueue(const AsyncQueue &other) = delete;
+        AsyncQueue &operator=(const AsyncQueue &other) = delete;
+
+        bool empty()
+        {
+            std::lock_guard<std::mutex> lock(_head_mutex);
+            return (_head.get() == _getTail());
+        }
+
+        std::shared_ptr<T> pop()
+        {
+            const auto old = _tryPop();
+            return old ? old->data : std::shared_ptr<T>();
+        }
+
+        bool pop(T &value)
+        {
+            const auto old = _tryPop(value);
+            return old != nullptr;
+        }
+
+        void push(T value)
+        {
+            auto data = std::make_shared<T>(std::move(value));
+            auto node = std::unique_ptr<Node>(new Node);
+            {
+                std::lock_guard<std::mutex> lock(_tail_mutex);
+                _tail->data = data;
+                Node *const new_tail = node.get();
+                _tail->next = std::move(node);
+                _tail = new_tail;
+            }
+            _condition.notify_one();
+        }
+
+      private:
+        struct Node {
+            std::shared_ptr<T> data;
+            std::unique_ptr<Node> next;
+        };
+
+        std::unique_ptr<Node> _head;
+        Node *_tail;
+
+        std::mutex _head_mutex;
+        std::mutex _tail_mutex;
+
+        std::condition_variable _condition;
+
+        Node *_getTail()
+        {
+            std::lock_guard<std::mutex> lock(_tail_mutex);
+            return _tail;
+        }
+
+        std::unique_ptr<Node> _pop()
+        {
+            if (_head.get() == _getTail()) {
+                return nullptr;
+            }
```

Comment:
This is also done is both of the methods that call it, so it is being done twice on each call to pop.  and the second _tryPop is also doing the same type of calls with the old head values.  Is this expected?